lib/coin_flip_demo.dart

```diff
@@ -0,0 +1,203 @@
+}
+
+class CoinFlipDemoState extends State<CoinFlipDemo> with SingleTickerProviderStateMixin {
+  AnimationController _controller;
+  Animation _flipAnimation;
+
+  // Whether the coin is currently flipping
+  bool isCoinFlipping = false;
+  // The number of times the coin flip resulted in heads.
+  int _headsCount = 0;
+  // The number of times the coin flip resulted in tails.
+  int _tailsCount = 0;
+  // The total number of times the coin has been flipped.
+  int _totalFlipCount = 0;
+
+  void _flipCoin() {
+    if (!isCoinFlipping) {
+      _headsCount = 0;
+      _tailsCount = 0;
+      _totalFlipCount = 250000;
+
+      setState(() {
+        isCoinFlipping = true;
+        for (int i = 0; i < _totalFlipCount; i++) {
+          bool isHeads = random.nextInt(2) % 2 == 0;
+          if (isHeads) {
+            _headsCount++;
+          } else {
+            _tailsCount++;
+          }
+        }
+        isCoinFlipping = false;
+      });
+    }
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = AnimationController(duration: Duration(milliseconds: 800), vsync: this);
+    final CurvedAnimation curve = CurvedAnimation(
+        parent: _controller,
+        curve: Interval(0.0, 1.0, curve: Curves.linear)
+    );
+    _flipAnimation = Tween(begin: 0.0, end: 1.0).animate(curve);
+    _controller.repeat();
+  }
+
+  dispose() {
+    _controller.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: Text('Coin Flip Demo'),
+      ),
+      body: Padding(
+        padding: EdgeInsets.all(8.0),
+        child: Column(
+          crossAxisAlignment: CrossAxisAlignment.center,
+          mainAxisAlignment: MainAxisAlignment.spaceAround,
+          children: <Widget>[
+            Column(
+              children: <Widget>[
+                _Coin(controller: _controller, flipAnimation: _flipAnimation),
+                SizedBox(height: 10.0),
+                _FlipButton(onPressed: _flipCoin)
+              ],
+            ),
+            _Results(heads: _headsCount, tails: _tailsCount, total: _totalFlipCount),
+          ],
+        ),
+      ),
+    );
+  }
+}
+
+class _Coin extends StatelessWidget {
+  _Coin({this.controller, this.flipAnimation});
+
+  final AnimationController controller;
+  final Animation flipAnimation;
+
+  @override
+  Widget build(BuildContext context) {
+    return Container(
+      padding: EdgeInsets.all(8.0),
+      child: AnimatedBuilder(
+        animation: controller,
+        builder: (BuildContext context, Widget child) {
+          return Center(
+            child: Container(
+              height: 150.0,
+              width: 150.0,
+              child: Transform(
+                transform: Matrix4.identity()
+                  ..rotateX(2 * pi * flipAnimation.value),
+                alignment: Alignment.center,
+                child: ClipOval(
```

Comment:
nit: to show best performance practices in this demo, it would be better to draw an oval directly than to use a ClipOval and a container.

Reply:
Done.